Source/WebCore/platform/graphics/PathSegmentData.cpp

```diff
@@ -0,0 +1,564 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "PathSegmentData.h"
+
+#include "AffineTransform.h"
+#include "PathImpl.h"
+#include <wtf/text/TextStream.h>
+
+namespace WebCore {
+
+FloatPoint PathMoveTo::calculateEndPoint(const FloatPoint&, FloatPoint& lastMoveToPoint) const
+{
+    lastMoveToPoint = point;
+    return lastMoveToPoint;
+}
+
+void PathMoveTo::extendFastBoundingRect(const FloatPoint&, FloatRect&) const
+{
+}
+
+void PathMoveTo::extendBoundingRect(const FloatPoint&, FloatRect&) const
+{
+}
+
+void PathMoveTo::addToImpl(PathImpl& impl) const
+{
+    impl.moveTo(point);
+}
+
+std::optional<PathElement> PathMoveTo::toElement() const
+{
+    return PathElement { PathElement::Type::MoveToPoint, { point } };
+}
+
+WTF::TextStream& operator<<(WTF::TextStream& ts, const PathMoveTo& data)
+{
+    ts << "move to " << data.point;
+    return ts;
+}
+
+FloatPoint PathLineTo::calculateEndPoint(const FloatPoint&, FloatPoint&) const
+{
+    return point;
+}
+
+void PathLineTo::extendFastBoundingRect(const FloatPoint& currentPoint, FloatRect& boundingRect) const
+{
+    extendBoundingRect(currentPoint, boundingRect);
+}
+
+void PathLineTo::extendBoundingRect(const FloatPoint& currentPoint, FloatRect& boundingRect) const
+{
+    boundingRect.extend(currentPoint);
+    boundingRect.extend(point);
+}
+
+void PathLineTo::addToImpl(PathImpl& impl) const
+{
+    impl.addLineTo(point);
+}
+
+std::optional<PathElement> PathLineTo::toElement() const
+{
+    return PathElement { PathElement::Type::AddLineToPoint, { point } };
+}
+
+WTF::TextStream& operator<<(WTF::TextStream& ts, const PathLineTo& data)
+{
+    ts << "add line to " << data.point;
+    return ts;
+}
+
+FloatPoint PathQuadCurveTo::calculateEndPoint(const FloatPoint&, FloatPoint&) const
+{
+    return endPoint;
+}
+
+void PathQuadCurveTo::extendFastBoundingRect(const FloatPoint& currentPoint, FloatRect& boundingRect) const
+{
+    boundingRect.extend(currentPoint);
+    boundingRect.extend(controlPoint);
+    boundingRect.extend(endPoint);
+}
+
+static float calculateQuadratic(float t, float p0, float p1, float p2)
```

Comment:
Since many of the calculations in this file are complex (particularly those for Beziers and arcs), it would be nice to see some tests. Maybe they could be API tests in Tools/TestWebKitAPI/Tests/WebCore/?